cpp/src/arrow/pretty_print.cc

```diff
@@ -413,18 +414,36 @@ Status ArrayPrinter::WriteValidityBitmap(const Array& array) {
   }
 }
 
+Result<std::shared_ptr<Array>> CopyStartEndToCPU(const Array& arr, int window) {
+  std::shared_ptr<Array> arr_sliced;
+  if (arr.length() > (2 * window + 1)) {
+    ARROW_ASSIGN_OR_RAISE(auto arr_start,
+                          arr.Slice(0, window + 1)->CopyTo(default_cpu_memory_manager()));
+    ARROW_ASSIGN_OR_RAISE(
+        auto arr_end,
+        arr.Slice(arr.length() - window - 1)->CopyTo(default_cpu_memory_manager()));
+    ARROW_ASSIGN_OR_RAISE(arr_sliced, Concatenate({arr_start, arr_end}));
+  } else {
+    ARROW_ASSIGN_OR_RAISE(arr_sliced, arr.CopyTo(default_cpu_memory_manager()));
```

Comment:
@pitrou I am using the slice+copy approach above in the idea of only copying a small part of the full array to the CPU device. However, on second though, I assume that this actually doesn't work as intended? The `Slice` is zero-copy (just adding offsets), and I assume the `CopyTo` just naively copies all buffers of the array in its entirety, not actually truncating and copying only the part of the buffer that is needed?

Reply:
Hmm, perhaps @felipecrv can answer these questions.

Reply:
> CopyTo just naively copies all buffers of the array in its entirety, not actually truncating and copying only the part of the buffer that is needed?

Yes, because `MemoryManager`s work in terms of Buffers, not ranges. We recently added a function that can copy slices to CPU, but it falls back to full buffer copy if the source `MemoryManager` is not a CPU memory manager.

The protocols for memory movement between devices might not allow for range-based copies (@zeroshade might know better than me), so instead expanding the `MemoryManager` interface we should probably have a way to slice the array where it is (device-specific) and then copy the smaller array. (?)

Reply:
Writing a slicer in CUDA might be a ton of work though.

Reply:
@felipecrv why does it fallback to a full buffer copy if the source `MemoryManager` isn't a CPU memory manager? Couldn't it just use `cuMemCpy` with the appropriate offset into the buffer?

Reply:
Oh, you're referring to slicing the *array* without copying the entirety of the buffers. Yea, that's a trickier proposition. The functions exist in libcudf already to do this, but we don't link against libcudf in libarrow_cuda (nor do I think we should), but in theory we *could* write the necessary code or at least figure it out by looking at how libcudf does it. For now, I think that the naive approach is likely fine until someone actively states they have an issue

Reply:
> For now, I think that the naive approach is likely fine until someone actively states they have an issue

So for the repr, there are basically two options:

- Don't print any actual data (eg in the Array repr, instead of including the data, add a general message like "xx values on xx device")
- Just do what I am doing here, acknowledging that the repr therefore _can_ be costly (if this is a problem, the user can still ensure to not actually use this / print the object). In practice, very often you will have chunked data, and so for example when having a table with many smaller chunks, we will still only copy a couple of chunks.

@zeroshade I take your comment as being fine with the second option?

Reply:
I'm fine with either approach. As long as we're able to indicate the situation without crashing, we're good. Acknowledging that the repr *can* be costly is a good tradeoff